### PR TITLE
[WIP] allow constant tensors in expressions

### LINKED ISSUE #7: Optimal path with constant arguments
I have an application where there is a large tensor contraction in the inner loop, with some arrays that are fixed and others that are updated between iterations. Operations only involving the constant terms could done ahead of time (outside the loop), but this may or may not actually improve performance.

It could be nice if opt-einsum had a way to mark some arguments as constant, in which case the cost of contracting them them could neglected when computing the cost of paths only involving these arguments.

### PR DESCRIPTION
**Pull request by jcmgray**
This allows constant tensors to be supplied to ``contract_expression``, which will then build as many constant intermediaries as it can. Still a bit of a work in progress but the core functionality is working so thought I'd share. Resolves #7.

## Example
Basic setup:
```python
>>> from opt_einsum import contract_expression
>>> from numpy.random import rand

>>> equation = 'ij,jk,kl,lm,mn->ni'
>>> const_ops = rand(2, 2), rand(2, 2), rand(2, 2)
>>> ops = (3, 2), *const_ops, (2, 3)  # non-const ops are just shapes

>>> expr = contract_expression(equation, *ops, constants=[1, 2, 3])
>>> expr
ContractExpression('ij,mn,[jk,kl,lm]->ni')

>>> print(expr)
<ContractExpression> for 'ij,mn,[jk,kl,lm]->ni':
  1.  'jm,ij->mi' [GEMM]
  2.  'mi,mn->ni' [GEMM]
```
Now we can call it with just two arrays:
```python
>>> expr(rand(3, 2), rand(2, 3))
array([[0.10982582, 0.06334146, 0.09379349],
       [0.09159049, 0.05282374, 0.07821902],
       [0.15775775, 0.09097925, 0.13471551]])

>>> expr(rand(3, 2), rand(2, 3))
array([[0.19207156, 0.21354127, 0.04811443],
       [0.01633236, 0.01815862, 0.004091  ],
       [0.06079124, 0.06758219, 0.0152304 ]])
```

## Notes
* Currently this just performs as many contractions in the ``contraction_list`` as possible, stopping when it reaches a non-constant array. This means that some potential constant contractions might still be missed even though they don't depend on anything (e.g. ``'a,a,b,b'`` with ``constants=[0, 1]`` might still choose to contract ``b,b`` first). One solution would be to somehow reshuffle the ``contraction_list`` but this gets a bit messy.
* How do we feel about the ``repr``?

## TODO
- [x] ~~Sort out interaction with ``backend`` (need to not cache expression on first constants pass, and deal with various conversion points (e.g. keep constants on GPU)~~
- [x] Document
- [ ] Add edge case tests? Additional errors to help user?

### CODE REVIEW

**Comment by dgasmith** [comment]
Cool! I will think a bit about how to observe constant expressions that are "future" in the path. Standard graph techniques could do this trivially, but seems like a lot of code when we can probably just use a few loops for it.

Also in the first example I think the constants `kwarg` should be `constants=[2, 3, 4]`?

**Comment by jcmgray** [comment]
> Cool! I will think a bit about how to observe constant expressions that are "future" in the path. Standard graph techniques could do this trivially, but seems like a lot of code when we can probably just use a few loops for it.

Yeah I haven't thought too much about these 'future' constant operations yet, but suspect they are quite rare.

One thing that would make addressing this easier would be to store the tensors not in a ``list`` but in a ``dict`` so that they are identifiable by more than just their position (which changes, meaning its hard to swap the order of operations). This kind of thing might also e.g. allow caching in the ``'optimal'`` algorithm and also reduce the cost of updating potential contractions in ``greedy``. Quite a big backend change though! And probably not necessary for this PR.

> Also in the first example I think the constants ``kwarg`` should be ``constants=[2, 3, 4]``?

The supplied ops are ``shape, array, array, array, shape`` so ``[1, 2, 3]`` is correct (if we are using 0-indexing). A point of confusion might be that the ``repr`` has all the constants moved to the end? :

```
'ij,jk,kl,lm,mn->ni' -> 'ij,mn,[jk,kl,lm]->ni'
```

**Comment by jcmgray** [comment]
Yes think I'll leave future constants for moment, as like you say, a more graph like structure will make them easier to implement.

Here's some themes and variations on potential representations:
```python
# 1 - nothing
"ContractExpression('ij,jk,kl,lm,mn->ni', constants=(1, 2, 4))"

# 2 - individually wrapped
"ContractExpression('ij,[jk,kl],lm,[mn]->ni', constants=(1, 2, 4))"

# 3 - Moved to end
"ContractExpression('ij,lm[jk,kl,mn]->ni')"

# 4 - explicitly mentioned in contraction string #1
"ContractExpression('ij,lm->ni [Constant: jk,kl,mn]')"

# 5 - explicitly mentioned in contraction string #2
"ContractExpression('ij,lm,[Const:jk,kl,mn]->ni')"

# 6 - explicitly mentioned in contraction string #3
"ContractExpression('ij,lm,Const[jk,kl,mn]->ni')"
```
I'd favour 2 or maybe 5. Its certainly nice to keep the them in order, apart from making it less clear how many arguments the expression should now be called with (realistically if the user has created the expression they probably know though!).

> Since we are moving around the order we may have to do something like Input: A_ij,B_[jk],C_[kl],D_[lm],E_mn->ni. Need to think about this more.

My 2c on introducing new letters is that might be a bit confusing, especially as the indices themselves could potentially be any upper or lower case characters.

---

I've also made a few tweaks to the backend stuff (and also will submit another PR later with support for torch and eager-mode tensorflow) but so far the performance gains from compiling these expressions with constants is pretty good, especially with a GPU!


> On switch to a dictionary it is certainly something to think about it. It would be nice to remove all of the indexing that we have floating around. I think you already switched the greedy algorithm to a dynamic approach to reduce the cost, not sure how it would directly effect optimal. It is certainly a point that we could do some sort of memoization on the contraction finder.

Yeah the logic for a dict in the optimal case is just that all tensors and contraction costs are uniquely defined by their indices (up to permutations), so a lot of different results could be reused regardless of however factorially many paths reach a certain point.

**Comment by dgasmith** [comment]
I guess my main concern is what if someone writes the following:

```
>>> contract_expression("ij,ij,ij,ij->, *ops, constants=[1, 3])
"ContractExpression('ij,ij,[Const:ij,ij]->')"
```

It is an edge case, but I think the loss of information there is pretty rough. All things considered I would vote for #2, I believe it is quite a bit more elegant than my try.

--

Great, I had been wondering about JIT for these expressions for a bit now. One thing to consider is `numba` or `numexpr` as well.

**Comment by jcmgray** [comment]
Yes that is a good edge case to point out! I've gone with number 2 as it does seem to be a good tradeoff.

I've also had a first run at some documentation. Let me know if anything could be clearer.

Was also thinking about adding this and ``opt_einsum`` 's other features to a kind of 'highlights' list on the main readme. Thoughts?  

----
> Great, I had been wondering about JIT for these expressions for a bit now. One thing to consider is numba or numexpr as well.

Yes - a recent ``numba`` release actually added support for proper n-dimensional reshaping and transposing of arrays which should mean a numba tensordot implementation is possible.

**Comment by dgasmith** [comment]
Adding to the readme and docs landing page would be good. However, I do worry a bit about overall length and the amount of content that we throw at people. I think these pages could be ~20-30% larger without too much issue. More than that is fine, but we would need to cut down on basic `opt_einsum` info in that case to keep the overall length rather short.

Overall this is looking great and is pretty close to being ready to go IMO.

File: opt_einsum/contract.py [hunk 1]
```diff
@@ -493,35 +498,109 @@ def _core_contract(operands, contraction_list, backend='numpy', **einsum_kwargs)
...
+        arrow = "->"
+    else:
+        lhs, rhs, arrow = einsum_str, "", ""
+
+    wrapped_terms = ["[{}]".format(t) if i in constants else t for i, t in enumerate(lhs.split(','))]
+
+    formatted_einsum_str = "{}{}{}".format(','.join(wrapped_terms), arrow, rhs)
+
+    # merge adjacent constants
+    formatted_einsum_str = formatted_einsum_str.replace("],[", ',')
```
[the remark this answers is not in the document]
Comment (jcmgray) [reply opt_einsum/contract.py]: Sometimes these things turn out easier than expected ;)

File: opt_einsum/contract.py [hunk 2]
```diff
@@ -438,9 +439,15 @@ def _core_contract(operands, contraction_list, backend='numpy', **einsum_kwargs)
...
         inds, idx_rm, einsum_str, remaining, blas_flag = contraction
-        tmp_operands = []
-        for x in inds:
-            tmp_operands.append(operands.pop(x))
+
+        # check if we are performing the pre-pass of an expression with constants
+        if parse_constants:
+            # if so, keep contracting until we find any non-constant operators, then break out
+            operands_copy = list(operands)
+            if any(operands_copy.pop(x) is None for x in inds):
```
[the remark this answers is not in the document]
Comment (jcmgray) [reply opt_einsum/contract.py]: I'll have a look

File: docs/source/reusing_paths.rst [hunk 3]
```diff
@@ -26,3 +26,73 @@ The ``ContractExpression`` can be called with 3 arrays that match the original s
...
+      2.  'jn,ij->ni' [GEMM]
+
+Where we can see that the expression now only has to perform
+two contractions to compute the output.
+
+.. note::
+
+    The constant part of an expression is lazily generated upon first call,
+    (with a particular backend) though it can be explicitly built with call to
+    :meth:`ContractExpression.parse_constants`.
```
[the remark this answers is not in the document]
Comment (jcmgray) [reply docs/source/reusing_paths.rst]: Yes I think that's a better name.

File: docs/source/reusing_paths.rst [hunk 4]
```diff
@@ -26,3 +26,73 @@ The ``ContractExpression`` can be called with 3 arrays that match the original s
...
+    >>> expr
+    <ContractExpression('ij,[jk,kl,lm],mn->ni', constants=[1, 2, 3])>
+
+The expression now only takes the remaining two arrays as arguments (the
+tensors with ``'ij'`` and ``'mn'`` indices), and will store as many resuable
+constant contractions as possible.
+
+.. code:: python
+
+    >>> out1 = expr(np.random.rand(9, 5), np.random.rand(5, 8))
```
[the remark this answers is not in the document]
Comment (jcmgray) [reply docs/source/reusing_paths.rst]: I'll add that in

File: opt_einsum/contract.py [hunk 5]
```diff
@@ -621,15 +721,29 @@ def contract_expression(subscripts, *shapes, **kwargs):
...
+        >>> a, b = np.random.rand(3, 4), np.random.rand(4, 5)
+        >>> c = expr(a, b)
+        >>> np.allclose(c, a @ b)
+        True
+
+    Supply ``a`` as a constant:
+
+        >>> expr = contract_expression("ab,bc->ac", a, (4, 5), constants=[0])
+        >>> expr
+        <ContractExpression('bc,[ab]->ac', constants=[0])>
```
Comment (dgasmith) [inline opt_einsum/contract.py]: This looks like an old version where the order was changed?

Comment (jcmgray) [reply opt_einsum/contract.py]: Yes good spot.

**Comment by jcmgray** [comment]
Thanks for the review comments, I'm just away at the moment but will try and make those changes soon. 

With regards to the README, I hear you about length. My thinking is just to give a hint of what is possible above and beyond numpy since some of the stuff added I assume is not going to be integrated there soon/ever. Will leave it to a later PR anyway.

**Comment by jcmgray** [comment]
OK, have finally made those updates @dgasmith

File: opt_einsum/contract.py [hunk 6]
```diff
@@ -438,9 +439,15 @@ def _core_contract(operands, contraction_list, backend='numpy', **einsum_kwargs)
...
         inds, idx_rm, einsum_str, remaining, blas_flag = contraction
-        tmp_operands = []
-        for x in inds:
-            tmp_operands.append(operands.pop(x))
+
+        # check if we are performing the pre-pass of an expression with constants
+        if evaluate_constants:
+            # if so, keep contracting until we find any non-constant operators, then break out
+            operands_copy = list(operands)
+            if any(operands_copy.pop(x) is None for x in inds):
```
Comment (dgasmith) [inline opt_einsum/contract.py]: I think my only comment is still here and if it is possible to convert this to set logic.

**Comment by dgasmith** [comment]
@jcmgray This all looks good to me. One small comment remaining, but nothing to block a merge. Is this ready to go?

File: opt_einsum/contract.py (hunk 6 above)
Comment (jcmgray) [reply opt_einsum/contract.py]: I'm not sure about set logic (if you have a snippet in mind please share!) but I did simplify this bit, based on realising that ``inds`` are actually always reverse-sorted so no popping needs to be done.

**Comment by jcmgray** [comment]
I've made a small improvement regarding that comment, but otherwise yes all good to go from my end.
